New Detector: TelerikLicenseKey [1039]

LanceMcCarthy opened the pull request:
### Description:
Added new detector for Telerik and Kendo License Keys (https://www.telerik.com). These JWTs are secrets which contain license information that must not be in source code.

This detector does everything locally by checking only the JWT's header for the `typ` attribute for our unique value "Telerik License Key".

#### Optimized Performance

There are no remote API calls to perform the detector's work, so I've also optimized the detector code and tests by removing httpClient usage for the following benefits:

✅ Reduced Dependencies, Cleaner Code, Better Performance: No HTTP client overhead (mocked or not)
✅ Simpler Testing: Tests are more focused on actual JWT validation logic
✅ All Tests Pass: Both unit tests and integration tests work perfectly (and the benchmarks look good too)

#### Tests

In my tests, there are two hard coded JWTs [secret](https://github.com/LanceMcCarthy/trufflehog/blob/d1e6fd90570d928503aaf5eaa5dd289c1065da7e/pkg/detectors/teleriklicensekey/teleriklicensekey_integration_test.go#L30C2-L30C8) and [inactiveSecret](https://github.com/LanceMcCarthy/trufflehog/blob/d1e6fd90570d928503aaf5eaa5dd289c1065da7e/pkg/detectors/teleriklicensekey/teleriklicensekey_integration_test.go#L32). _This is nothing to be concerned about_, they are sample values generated specifically for long term testing support and are **not** secrets. 

One has the expected header value to verify a confirmed positive hit, while the other is like any other normal JWT, and does not trigger a false positive.

<img width="960" height="734" alt="image" src="https://github.com/user-attachments/assets/52cabbcc-a8d1-45b6-85bf-05bc7bde7ff6" />

> [!Note]
> If these test keys trigger your secret-detection workflow, please let me know how to add an ignore for them. Alternatively, I can switch back to the GCP approach, but that was giving me nothing but headaches.

### Checklist:

* ✅ Tests passing (`make test-community`)?
    * <img width="766" height="223" alt="image" src="https://github.com/user-attachments/assets/60ddad08-d694-4212-bbfa-33c28bb763a8" />
* [ ] Lint passing (`make lint` this requires [golangci-lint](https://golangci-lint.run/welcome/install/#local-installation))?
    * I attempted to run this, but it seems to be an outdated instruction as there's a default flag not accepted. Might be outdated tooling in the GitHub Codespace, I can rerun with additional instructions.
    * <img width="884" height="85" alt="image" src="https://github.com/user-attachments/assets/72069905-11c5-4b14-90dc-9abcac35fca6" />

[EDIT] Spelling corrections, and added links/note about the test keys.

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> **Medium Risk**
> Adds a new detector type and wires it into default detector registration and protobuf enums, which can affect detector selection/serialization and overall scan behavior. Detection logic is local-only but still introduces new matching patterns that could impact performance/false positives.
> 
> **Overview**
> Adds a new `TelerikLicenseKey` detector that identifies Telerik/Kendo license JWTs by regex-matching JWT-shaped tokens and locally validating the decoded header `typ` claim.
> 
> Registers the detector in the default detector list and introduces a new protobuf `DetectorType_TelerikLicenseKey` enum value, plus config parsing/string tests to ensure the new detector can be selected by name and rendered consistently.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit bd0666f60d053157eaf146e35249b42ab14108ec. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

LanceMcCarthy committed:
Added TelerikLicenseKey Detector ID

LanceMcCarthy committed:
Generated  default detector code for TelerikLicenseKey detector

LanceMcCarthy committed:
Added regex for detecting Telerik Licensing JWTs.

LanceMcCarthy committed:
Update detector description

LanceMcCarthy committed:
Set testing keys

LanceMcCarthy committed:
Decoding and JWT header check added, ongoing test rewrite.

LanceMcCarthy committed:
In progress

LanceMcCarthy committed:
Finished tests and fixed all verification/validation logic for local Telerik JWT decoding/validation

LanceMcCarthy committed:
Removed HttpClient dep/usage

✅ Reduced Dependencies: No longer depends on HTTP client libraries
✅ Cleaner Code: Removed unused HTTP client logic
✅ Better Performance: No HTTP client initialization overhead
✅ Simpler Testing: Tests are more focused on actual JWT validation logic
✅ All Tests Pass: Both unit tests and integration tests work perfectly

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/trufflesecurity/trufflehog?pullRequest=4479) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/trufflesecurity/trufflehog?pullRequest=4479) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/trufflesecurity/trufflehog?pullRequest=4479) it.</sub>

LanceMcCarthy commented:
Thank you for the prompt follow up and answers @shahzadhaider1! I was eagerly awaiting the results of the checks because I couldn't run the linter manually, but the two linter workflows/check look good :)

My remaining worry was if the 'scan for secrets` check would throw a false positive because of my test JWTs. Is there a reason they were skipped, do I need the code review first?

LanceMcCarthy commented:
> > Hi @shahzadhaider1 I wanted to check in with you again. Is there anything I can do to help speed tings up? Is there something I missed?
> 
> Hey @LanceMcCarthy, we’re currently pausing the addition of new detectors. We’ll get this merged once we’ve refined our process for maintaining them.

Ah! Okay, thanks for letting me know that the delay is not due to something on my side.

LanceMcCarthy committed:
Address both Cursor BugBot issues

This resolves two items raised in the PR

- ☑️ Missing MaxSecretSizeProvider causes truncation of long JWTs
- ☑️ Protobuf name/value maps missing TelerikLicenseKey entry
- Additional refinements.

LanceMcCarthy committed:
Address new Bugbot items

Resolved:

- Regex character class has unintended range operator
- Detector not registered in engine defaults

LanceMcCarthy committed:
More Bugbot fixes

LanceMcCarthy committed:
More Bugbot rsponses

Can someone change the Bugbot setting so that it can give you more than 2 at time to fix?

LanceMcCarthy committed:
One more bugbot item to address

LanceMcCarthy committed:
Update tests

LanceMcCarthy committed:
Aaaand one mor Bugbot follow up (can someone change the setting to allow Bugbot to report all unique items at once)

LanceMcCarthy commented:
Okay after the Bugbot review and adjustments, @shahzadhaider1 it's ready :D

IF I had any feedback for you team, see if Bugbot has a setting that allows for more than 2 issues at once. While yes, I understand newly introduced issue, but this was 95% existing issues that could have bene listed in a single recommendation flow

LanceMcCarthy committed:
Merge remote-tracking branch 'upstream/main' into prgs-devtools-license-key-detector

LanceMcCarthy committed:
Migrated detector to use new detector_type.pb approach

LanceMcCarthy commented:
@shahzadhaider1 can you please review this again. I have 

Pulled in all the changes from main
Completely migrated my detector to the new approach

Everything lines up to how you need it, you can squash this into a single commit and it will line up perfectly.

Here's the latest test result:

```bash
go test -v ./pkg/config ./pkg/detectors/teleriklicensekey -tags=detectors

=== RUN   TestDetectorParsing
...
=== RUN   TestDetectorParsing/new_detector_by_name
...
--- PASS: TestDetectorParsing (0.00s)
...
=== RUN   TestDetectorIDString
--- PASS: TestDetectorIDString (0.00s)
PASS
ok      github.com/trufflesecurity/trufflehog/v3/pkg/config     4.047s

=== RUN   TestTeleriklicensekey_FromChunk
=== RUN   TestTeleriklicensekey_FromChunk/found,_not_verified
=== RUN   TestTeleriklicensekey_FromChunk/non-telerik_jwt_is_filtered
=== RUN   TestTeleriklicensekey_FromChunk/not_found
--- PASS: TestTeleriklicensekey_FromChunk (0.00s)
...
=== RUN   TestTeleriklicensekey_Pattern
=== RUN   TestTeleriklicensekey_Pattern/Valid_JWT_-_Telerik_License_Key
=== RUN   TestTeleriklicensekey_Pattern/Malformed_JWT_-_no_JWT_pattern_detected
=== RUN   TestTeleriklicensekey_Pattern/No_JWT_present
--- PASS: TestTeleriklicensekey_Pattern (0.00s)
=== RUN   TestTeleriklicensekey_LongJWTChunkSpan
--- PASS: TestTeleriklicensekey_LongJWTChunkSpan (0.00s)
PASS
ok      github.com/trufflesecurity/trufflehog/v3/pkg/detectors/teleriklicensekey 2.828s
```

shahzadhaider1 commented:
> @shahzadhaider1 can you please review this again. I have
> 
> Pulled in all the changes from main Completely migrated my detector to the new approach
> 
> Everything lines up to how you need it, you can squash this into a single commit and it will line up perfectly.
> 
> Here's the latest test result:
> 
> ```shell
> go test -v ./pkg/config ./pkg/detectors/teleriklicensekey -tags=detectors
> 
> === RUN   TestDetectorParsing
> ...
> === RUN   TestDetectorParsing/new_detector_by_name
> ...
> --- PASS: TestDetectorParsing (0.00s)
> ...
> === RUN   TestDetectorIDString
> --- PASS: TestDetectorIDString (0.00s)
> PASS
> ok      github.com/trufflesecurity/trufflehog/v3/pkg/config     4.047s
> 
> === RUN   TestTeleriklicensekey_FromChunk
> === RUN   TestTeleriklicensekey_FromChunk/found,_not_verified
> === RUN   TestTeleriklicensekey_FromChunk/non-telerik_jwt_is_filtered
> === RUN   TestTeleriklicensekey_FromChunk/not_found
> --- PASS: TestTeleriklicensekey_FromChunk (0.00s)
> ...
> === RUN   TestTeleriklicensekey_Pattern
> === RUN   TestTeleriklicensekey_Pattern/Valid_JWT_-_Telerik_License_Key
> === RUN   TestTeleriklicensekey_Pattern/Malformed_JWT_-_no_JWT_pattern_detected
> === RUN   TestTeleriklicensekey_Pattern/No_JWT_present
> --- PASS: TestTeleriklicensekey_Pattern (0.00s)
> === RUN   TestTeleriklicensekey_LongJWTChunkSpan
> --- PASS: TestTeleriklicensekey_LongJWTChunkSpan (0.00s)
> PASS
> ok      github.com/trufflesecurity/trufflehog/v3/pkg/detectors/teleriklicensekey 2.828s
> ```

Hey, thanks for your patience--this is solid work and the implementation looks good overall.
I do have one concern though: the regex pattern used here appears to overlap with the existing JWT detector in TruffleHog. I did a quick local test with the latest version, using the sample secret from your test, and it was already detected by the JWT detector.

I might be missing something, but do you think this detector adds meaningful value given that similar secrets are already being caught? Would be great to understand if there are cases this would uniquely cover.

amanfcp commented:
Hey @LanceMcCarthy, thanks for the contribution. 

We’re currently prioritizing updates to high-usage/high-impact detectors. Other detector PRs are being queued and will be reviewed and merged based on their relative impact.

We’re also being deliberate with detector additions/updates to maintain signal quality and avoid regressions or anomalous behavior observed in prior changes.

LanceMcCarthy commented:
Hi team!

@shahzadhaider1 
I will review it again, but the last time I checked, the default TruffleHog JWT detector did not match our license keys. Today, I will run a test against a large amount of sample keys niu another rpo, using the latest release of trufflehog, then I'll come back here and share the results.

@amanfcp 
Understood, thank you. To be clear, I'm not proposing this is a high-priority opsec detector. Rather it is an infosec detector, it is important for Progress Software and all of our customers; they don't want to leak their license keys, and it is a very common mistake on their part.

Thank you both for taking the time to discuss, I sincerely appreciate it. Shahzad, I will get back to you with the results of my test.

LanceMcCarthy committed:
Merge remote-tracking branch 'upstream/main' into prgs-devtools-license-key-detector

LanceMcCarthy committed:
Gen Protos

LanceMcCarthy committed:
Update teleriklicensekey.go

LanceMcCarthy committed:
Add verified flag

LanceMcCarthy commented:
Hi team @amanfcp ,

I have some updates to share, see the following screenshot where I run this command against a set of sample license keys

`docker run --rm -it -v "${PWD}:/pwd" trufflesecurity/trufflehog:latest filesystem /pwd`

<img width="900" alt="image" src="https://github.com/user-attachments/assets/65a87f0d-323f-4bef-a30d-4cb712a3e2b5" />

It looks like it is now finding license keys, likely due to the updates you did for this announcement 3 months after I opened this PR https://trufflesecurity.com/blog/trufflehog-now-detects-jwts-with-public-key-signatures-and-verifies-them-for-liveness

However, none of them are verified. So, I have updated this PR so that it matches Telerik license keys. Here's the tests

<img width="740" alt="image" src="https://github.com/user-attachments/assets/5df888e0-e74a-4ed8-be2f-2f8d644113ab" />

in case you want to merge it. Otherwise, feel free to close this PR
